File: src/c/framework/src/world.c
// ke_world impl — default framework aggregator. Owns ecs+runtime+scene_tree
// transferred from host on create; cascades teardown in reverse order on destroy.

// scene_tree.h is included BEFORE world_create.h so that its transitive
// `framework_export.h` define of KE_FRAMEWORK_API runs first; world_create.h's
// own fallback definition then sees the macro already defined and skips it.
#include <kernel_engine/framework/scene_tree.h>
#include <kernel_engine/framework/components.h>
#include <kernel_engine/framework/world_create.h>
#include <kernel_engine/common/error.h>
#include <kernel_engine/allocator/allocator.h>
#include "components_apply.h"

#include <stddef.h>
#include <string.h>

// Component-apply registry: a tiny linear array of (cid, fn) pairs. The
// scene_loader is the only consumer today; lookups happen at scene-load time
// (not per-frame), so linear scan over the typical 6-20 registered apply
// entries is fine. If the registry grows past ~64 we can swap for a hash.
typedef struct apply_entry {
    ke_component_id        cid;
    ke_component_apply_fn  fn;
} apply_entry;

typedef struct ke_world_state {
    ke_allocator             *allocator;       // owned — created internally in ke_world_create
    struct ke_task_scheduler *task_scheduler;  // borrowed
    ke_ecs                   *ecs;             // owned
    ke_runtime               *runtime;         // owned
    struct ke_scene_tree     *scene_tree;      // owned (NULL allowed during C-phase transition)
    const char               *project_root;    // borrowed string
    struct ke_logger         *logger;          // borrowed

    apply_entry              *apply_registry;
    uint32_t                  apply_count;
    uint32_t                  apply_capacity;
} ke_world_state;

static ke_ecs *world_ecs(struct ke_world *self) {
    ke_world_state *s = (ke_world_state *)self->handle;
    return s->ecs;
}

static ke_runtime *world_runtime(struct ke_world *self) {
    ke_world_state *s = (ke_world_state *)self->handle;
    return s->runtime;
}

static struct ke_scene_tree *world_scene_tree(struct ke_world *self) {
    ke_world_state *s = (ke_world_state *)self->handle;
    return s->scene_tree;
}

static ke_result world_register_component_apply(struct ke_world *self,
                                                  ke_component_id cid,
                                                  ke_component_apply_fn apply,
                                                  ke_error **out_error) {
    if (!self || !self->handle || !apply) return KE_ERROR_SET(out_error, &KE_ERROR_INVALID_ARGUMENT, "invalid argument");
    ke_world_state *s = (ke_world_state *)self->handle;

    // Replace-if-exists: registering the same cid twice updates the function.
    for (uint32_t i = 0; i < s->apply_count; ++i) {
        if (s->apply_registry[i].cid == cid) {
            s->apply_registry[i].fn = apply;
            return KE_OK;
        }
    }

    if (s->apply_count == s->apply_capacity) {
        uint32_t cap = s->apply_capacity ? s->apply_capacity * 2 : 16;
        apply_entry *new_buf = (apply_entry *)s->allocator->alloc(
            s->allocator, sizeof(apply_entry) * cap, 8);
        if (!new_buf) return KE_ERROR_SET(out_error, &KE_ERROR_OUT_OF_MEMORY, "apply registry allocation failed");
        if (s->apply_registry) {
            memcpy(new_buf, s->apply_registry, sizeof(apply_entry) * s->apply_count);
            s->allocator->free(s->allocator, s->apply_registry);
        }
        s->apply_registry = new_buf;
        s->apply_capacity = cap;
    }
    s->apply_registry[s->apply_count].cid = cid;
    s->apply_registry[s->apply_count].fn  = apply;
    s->apply_count++;
    return KE_OK;
}

static ke_component_apply_fn world_get_component_apply(struct ke_world *self,
                                                        ke_component_id cid) {
    if (!self || !self->handle) return NULL;
    ke_world_state *s = (ke_world_state *)self->handle;
    for (uint32_t i = 0; i < s->apply_count; ++i) {
        if (s->apply_registry[i].cid == cid) return s->apply_registry[i].fn;
    }
    return NULL;
}

static void world_destroy(struct ke_world *self) {
    if (!self) return;
    ke_world_state *s = (ke_world_state *)self->handle;
    if (s) {
        if (s->apply_registry) s->allocator->free(s->allocator, s->apply_registry);

        // ecs, runtime, scene_tree are BORROWED — caller destroys them after world->destroy().
        // "quem cria, owna": world did not create these; world must not destroy them.
        ke_allocator *a = s->allocator;
        a->free(a, s);
        a->destroy(a);
    }
    // self lives in the same allocation as state — already freed.
}

ke_result ke_world_create(const ke_world_params *params, ke_world **out_world, ke_error **out_error) {
    if (!params || !out_world) return KE_ERROR_SET(out_error, &KE_ERROR_INVALID_ARGUMENT, "invalid argument");
    if (!params->ecs || !params->runtime) {
        return KE_ERROR_SET(out_error, &KE_ERROR_INVALID_ARGUMENT, "ecs and runtime are required");
    }

    ke_allocator *a = ke_allocator_malloc_create();
    if (!a) return KE_ERROR_SET(out_error, &KE_ERROR_OUT_OF_MEMORY, "allocator creation failed");

    // Single allocation: state + vtable contiguous. Simpler teardown.
    size_t block_size = sizeof(ke_world_state) + sizeof(ke_world);
    void *block = a->alloc(a, block_size, 8);
    if (!block) { a->destroy(a); return KE_ERROR_SET(out_error, &KE_ERROR_OUT_OF_MEMORY, "state allocation failed"); }
    memset(block, 0, block_size);

    ke_world_state *state = (ke_world_state *)block;
    ke_world       *world = (ke_world *)((char *)block + sizeof(ke_world_state));

    state->allocator      = a;
    state->task_scheduler = params->task_scheduler;
    state->ecs            = params->ecs;
    state->runtime        = params->runtime;
    state->scene_tree     = params->scene_tree;
    state->project_root   = params->project_root;
    state->logger         = params->logger;

    world->handle                  = state;
    world->ecs                     = world_ecs;
    world->runtime                 = world_runtime;
    world->scene_tree              = world_scene_tree;
    world->register_component_apply = world_register_component_apply;
    world->get_component_apply     = world_get_component_apply;
    world->destroy                 = world_destroy;

    // Register the framework's built-in component vocabulary with the ecs +
    // wire up each component's apply callback. component_register is
    // idempotent in spirit — scene_tree already registers transform/hierarchy/
    // name via component_lookup-first; the others are first-touch here. The
    // host pays the schema setup cost once per world; the scene_loader then
    // works out of the box for any of these components.
    ke_ecs *e = state->ecs;
    ke_component_meta meta;
    #define REG(name, type, apply_fn) do { \
        ke_component_id _cid; \
        if (e->component_lookup(e, (name), &meta, NULL) == KE_OK) { _cid = meta.cid; } \
        else { _cid = e->component_register(e, (name), sizeof(type)); } \
        world->register_component_apply(world, _cid, (apply_fn), NULL); \
    } while (0)

    REG(KE_COMPONENT_NAME_TRANSFORM,         ke_transform_component,        ke_framework_apply_transform);
    REG(KE_COMPONENT_NAME_CAMERA,            ke_camera_component,           ke_framework_apply_camera);
    REG(KE_COMPONENT_NAME_MESH,              ke_mesh_component,             ke_framework_apply_mesh);
    REG(KE_COMPONENT_NAME_DIRECTIONAL_LIGHT, ke_directional_light_component, ke_framework_apply_directional_light);
    REG(KE_COMPONENT_NAME_POINT_LIGHT,       ke_point_light_component,      ke_framework_apply_point_light);
    REG(KE_COMPONENT_NAME_SPOT_LIGHT,        ke_spot_light_component,       ke_framework_apply_spot_light);
    #undef REG

    *out_world = world;
    return KE_OK;
}

File: src/zig/framework/src/world.zig
// ke_world impl — default framework aggregator. Holds ecs+runtime+scene_tree
// handed over by the host on create, and the component-apply registry the
// scene loader consults.

const std = @import("std");

const c = @import("c.zig").c;
const heap = @import("heap.zig");

const E = @import("kerror").Errors(c);

const apply = @import("components_apply.zig");

/// Starting size of the apply registry; it doubles on demand, so this only
/// trades a little memory against a few early reallocs.
const apply_initial_capacity: u32 = 16;

// A tiny linear array of (cid, fn) pairs. The scene loader is the only consumer
// today and lookups happen at scene-load time, not per frame, so a linear scan
// over the typical 6-20 entries is fine. Past ~64 this wants a hash.
const ApplyEntry = struct {
    cid: c.ke_component_id,
    fn_ptr: c.ke_component_apply_fn,
};

const State = struct {
    scheduler: ?*c.ke_scheduler, // borrowed
    ecs: ?*c.ke_ecs, // borrowed
    runtime: ?*c.ke_runtime, // borrowed
    scene_tree: ?*c.ke_scene_tree, // borrowed
    project_root: [*c]const u8, // borrowed string
    logger: ?*c.ke_logger, // borrowed

    apply_registry: ?[*]ApplyEntry,
    apply_count: u32,
    apply_capacity: u32,
};

// State and vtable live in one allocation, so freeing the state frees both.
const Block = struct {
    state: State,
    world: c.ke_world,
};

fn stateOf(self: *c.ke_world) *State {
    return @ptrCast(@alignCast(self.handle));
}

fn worldEcs(self_in: ?*c.ke_world) callconv(.c) ?*c.ke_ecs {
    const self = self_in orelse return null;
    return stateOf(self).ecs;
}

fn worldRuntime(self_in: ?*c.ke_world) callconv(.c) ?*c.ke_runtime {
    const self = self_in orelse return null;
    return stateOf(self).runtime;
}

fn worldSceneTree(self_in: ?*c.ke_world) callconv(.c) ?*c.ke_scene_tree {
    const self = self_in orelse return null;
    return stateOf(self).scene_tree;
}

fn worldRegisterComponentApply(
    self_in: ?*c.ke_world,
    cid: c.ke_component_id,
    apply_fn: c.ke_component_apply_fn,
    out_error: [*c][*c]c.ke_error,
) callconv(.c) bool {
    const self = self_in orelse {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return false;
    };
    if (self.handle == null or apply_fn == null) {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return false;
    }
    const s = stateOf(self);

    // Replace-if-exists: registering the same cid twice updates the function.
    if (s.apply_registry) |reg| {
        for (reg[0..s.apply_count]) |*entry| {
            if (entry.cid == cid) {
                entry.fn_ptr = apply_fn;
                return true;
            }
        }
    }

    if (s.apply_count == s.apply_capacity) {
        const cap = if (s.apply_capacity != 0) s.apply_capacity * 2 else apply_initial_capacity;
        const new_buf = heap.gpa.alloc(ApplyEntry, cap) catch {
            E.fail(out_error, .out_of_memory, "apply registry allocation failed", @src());
            return false;
        };
        if (s.apply_registry) |old| {
            @memcpy(new_buf[0..s.apply_count], old[0..s.apply_count]);
            // Released at the capacity it was allocated with, which the state
            // still holds until the new one is published below.
            heap.gpa.free(old[0..s.apply_capacity]);
        }
        s.apply_registry = new_buf.ptr;
        s.apply_capacity = cap;
    }

    s.apply_registry.?[s.apply_count] = .{ .cid = cid, .fn_ptr = apply_fn };
    s.apply_count += 1;
    return true;
}

fn worldGetComponentApply(
    self_in: ?*c.ke_world,
    cid: c.ke_component_id,
) callconv(.c) c.ke_component_apply_fn {
    const self = self_in orelse return null;
    if (self.handle == null) return null;
    const s = stateOf(self);
    const reg = s.apply_registry orelse return null;
    for (reg[0..s.apply_count]) |entry| {
        if (entry.cid == cid) return entry.fn_ptr;
    }
    return null;
}

fn worldDestroy(self_in: ?*c.ke_world) callconv(.c) void {
    const self = self_in orelse return;
    if (self.handle == null) return;
    const s = stateOf(self);
    if (s.apply_registry) |reg| heap.gpa.free(reg[0..s.apply_capacity]);
    // ecs, runtime and scene_tree are borrowed: whoever created them destroys
    // them after world->destroy(). The state sits at the head of the block, so
    // freeing it releases the vtable too.
    heap.gpa.destroy(@as(*Block, @fieldParentPtr("state", s)));
}

/// Registers a built-in component with the ecs (reusing an existing
/// registration when the name is already known) and wires up its apply callback.
fn registerBuiltin(
    world: *c.ke_world,
    e: *c.ke_ecs,
    name: [*c]const u8,
    size: usize,
    apply_fn: c.ke_component_apply_fn,
) void {
    var meta: c.ke_component_meta = undefined;
    const cid = if (e.component_lookup.?(e, name, &meta, null))
        meta.cid
    else
        e.component_register.?(e, name, size);
    _ = world.register_component_apply.?(world, cid, apply_fn, null);
}

export fn ke_world_create(
    params_in: ?*const c.ke_world_params,
    out_error: [*c][*c]c.ke_error,
) callconv(.c) c.ke_world_handle {
    const null_handle = std.mem.zeroes(c.ke_world_handle);

    const params = params_in orelse {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return null_handle;
    };
    if (params.ecs == null or params.runtime == null) {
        E.fail(out_error, .invalid_argument, "ecs and runtime are required", @src());
        return null_handle;
    }

    const block = heap.gpa.create(Block) catch {
        E.fail(out_error, .out_of_memory, "state allocation failed", @src());
        return null_handle;
    };
    block.* = std.mem.zeroes(Block);

    const state = &block.state;
    const world = &block.world;

    state.scheduler = params.scheduler;
    state.ecs = params.ecs;
    state.runtime = params.runtime;
    state.scene_tree = params.scene_tree;
    state.project_root = params.project_root;
    state.logger = params.logger;

    world.handle = state;
    world.ecs = worldEcs;
    world.runtime = worldRuntime;
    world.scene_tree = worldSceneTree;
    world.register_component_apply = worldRegisterComponentApply;
    world.get_component_apply = worldGetComponentApply;

    // Register the framework's built-in component vocabulary and wire each
    // component's apply callback, so the scene loader works out of the box.
    // scene_tree already registers transform/hierarchy/name; the rest are
    // first-touch here. The host pays this schema setup once per world.
    const e = params.ecs.?;
    registerBuiltin(world, e, c.KE_COMPONENT_NAME_TRANSFORM, @sizeOf(c.ke_transform_component), apply.ke_framework_apply_transform);
    registerBuiltin(world, e, c.KE_COMPONENT_NAME_CAMERA, @sizeOf(c.ke_camera_component), apply.ke_framework_apply_camera);
    registerBuiltin(world, e, c.KE_COMPONENT_NAME_MESH, @sizeOf(c.ke_mesh_component), apply.ke_framework_apply_mesh);
    registerBuiltin(world, e, c.KE_COMPONENT_NAME_DIRECTIONAL_LIGHT, @sizeOf(c.ke_directional_light_component), apply.ke_framework_apply_directional_light);
    registerBuiltin(world, e, c.KE_COMPONENT_NAME_POINT_LIGHT, @sizeOf(c.ke_point_light_component), apply.ke_framework_apply_point_light);
    registerBuiltin(world, e, c.KE_COMPONENT_NAME_SPOT_LIGHT, @sizeOf(c.ke_spot_light_component), apply.ke_framework_apply_spot_light);

    return .{ .ref = world, .destroy = worldDestroy };
}
